### PR TITLE
Made checkbox controlled

### LINKED ISSUE #2970: Make Checkbox component controllable
Currently, the `CheckBox` component is designed as an uncontrolled component, meaning its checked state is managed internally within the component using useState. This makes it impossible to control the checkbox state from outside the component.
**Expected Behavior:** We need to refactor the `CheckBox` component to support both controlled and uncontrolled states. The state should be able to be managed externally through props (e.g., `checked`, `onChange`) and internally as a fallback when the checked prop is not provided.

`Proposed Solution:`

- Add a `checked` prop to the component to allow external control of the checkbox's checked state.
- Add an `onChange` callback prop that will be triggered when the checkbox state changes.
- If the `checked` prop is provided, use it as the source of truth for the checkbox state; otherwise, fall back to using the internal state (with useState).
- Ensure the component still works correctly when used in an uncontrolled manner by managing the state internally when `checked` is not provided.
- Update storybook to reflect these changes

### PR DESCRIPTION
**Pull request by nebby2105**
Refactored the CheckBox component to support both controlled and uncontrolled states. #2970 
The changes allow external state management via props (`checked` and `onChange`) while preserving internal state handling as a fallback when the `checked` prop is not provided.

Here’s a video demonstrating how the controlled CheckBox interacts with external state:

https://github.com/user-attachments/assets/359751f5-0512-4eb9-b9e9-f31b599ef41e

Summary:

- [x] Added a `checked` prop to enable external control of the checkbox state.
- [x] Introduced an `onChange` callback prop to handle state change events.
- [x] Implemented fallback logic to use internal state when the `checked` prop is not passed.
- [x] Preserved support for uncontrolled usage with the `defaultChecked` prop.
- [x] Added a new "Controlled" story to demonstrate the controlled behavior of the CheckBox component.
- [x] Updated component descriptions to reflect support for both controlled and uncontrolled states.

### CODE REVIEW

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 1]
```diff
@@ -12,9 +12,16 @@ interface CheckBoxProps extends Omit<CheckboxProps, 'size'> {
   color?: 'primary' | 'secondary' | 'error' | 'success'
   size?: 'sm' | 'md' | 'lg'
   loading?: boolean
+  checked?: boolean
+  onChange?: (
+    event: React.ChangeEvent<HTMLInputElement>,
```
Comment (ShadowOfTheSpace) [inline src/design-system/components/checkbox/CheckBox.tsx]: We don't need to pass this event to the `onChange` callback

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 2]
```diff
@@ -24,10 +31,18 @@ const CheckBox: FC<CheckBoxProps> = ({
   size = 'md',
   ...props
 }) => {
-  const [checked, setChecked] = useState<boolean>(false)
+  const [internalChecked, setInternalChecked] = useState<boolean>(false)
```
Comment (ShadowOfTheSpace) [inline src/design-system/components/checkbox/CheckBox.tsx]: Add `defaultIsChecked` prop and use it here. And also rename state to `internalIsChecked`, please
```suggestion
  const [internalChecked, setInternalChecked] = useState<boolean>(defaultIsChecked ?? false)
```

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 3]
```diff
@@ -12,9 +12,16 @@ interface CheckBoxProps extends Omit<CheckboxProps, 'size'> {
...
   loading?: boolean
+  checked?: boolean
+  onChange?: (
+    event: React.ChangeEvent<HTMLInputElement>,
+    checked: boolean
+  ) => void
 }
 
 const CheckBox: FC<CheckBoxProps> = ({
+  checked,
```
Comment (ShadowOfTheSpace) [inline src/design-system/components/checkbox/CheckBox.tsx]: `checked` is a boolean, so rename it to `isChecked`. But also, to be more consistent in naming in this component, you should use rename while destructuring.
So you will have:
```isChecked: externalIsChecked```

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 4]
```diff
@@ -24,10 +31,18 @@ const CheckBox: FC<CheckBoxProps> = ({
...
   ...props
 }) => {
-  const [checked, setChecked] = useState<boolean>(false)
+  const [internalChecked, setInternalChecked] = useState<boolean>(false)
+
+  const isCheckboxControllable = checked !== undefined
 
   const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
-    setChecked(event.target.checked)
+    const newChecked = event.target.checked
```
Comment (ShadowOfTheSpace) [inline src/design-system/components/checkbox/CheckBox.tsx]: We don't need event in `onChange` callback, so you can use destructurization here or even in function declaration, like `{ target: { checked } } : React.ChangeEvent<HTMLInputElement>)`

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 5]
```diff
@@ -37,6 +52,7 @@ const CheckBox: FC<CheckBoxProps> = ({
   }
 
   const loader = <Loader size={loaderSizeMapping[size]} />
+  const resolvedCheck = isCheckboxControllable ? checked : internalChecked
```
Comment (ShadowOfTheSpace) [inline src/design-system/components/checkbox/CheckBox.tsx]:
```suggestion
  const resolvedIsChecked = isCheckboxControllable ? externalIsChecked: internalIsChecked
```

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 6]
```diff
@@ -60,12 +76,12 @@ const CheckBox: FC<CheckBoxProps> = ({
       ) : (
         <Checkbox
           {...props}
-          checked={checked}
+          checked={resolvedCheck}
```
Comment (ShadowOfTheSpace) [inline src/design-system/components/checkbox/CheckBox.tsx]:
```suggestion
          checked={resolvedIsChecked}
```

**Review by ShadowOfTheSpace — CHANGES_REQUESTED** [review]

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 7]
```diff
@@ -1,20 +1,24 @@
 import Checkbox, { CheckboxProps } from '@mui/material/Checkbox'
-import { FC, ReactNode, useState } from 'react'
+import { type ChangeEvent, ReactNode, useCallback, useState } from 'react'
```
Comment (ShadowOfTheSpace) [inline src/design-system/components/checkbox/CheckBox.tsx]:
```suggestion
import { type ChangeEvent, type ReactNode, useCallback, useState } from 'react'
```

**Review by ShadowOfTheSpace — CHANGES_REQUESTED** [review]

File: src/design-system/components/checkbox/CheckBox.tsx [hunk 8]
```diff
@@ -24,11 +30,22 @@ const CheckBox: FC<CheckBoxProps> = ({
   size = 'md',
   ...props
 }) => {
-  const [checked, setChecked] = useState<boolean>(false)
+  const [internalIsChecked, setInternalIsChecked] = useState<boolean>(
+    defaultChecked ?? false
```
Comment (ShadowOfTheSpace) [inline src/design-system/components/checkbox/CheckBox.tsx]: You already assigned default value to the `defaultChecked` prop

**Review by ShadowOfTheSpace — APPROVED** [review]

**Review by luiqor — APPROVED** [review]

**Review by markgol777 — APPROVED** [review]

**Pull request merged**